### PR TITLE
Automatic Test Generation

### PR DESCRIPTION
**Pull request by stevenpeniche**
Newton's Apple will generate a test when creating a new component if the `-t` or `--test` options are passed. Furthermore. automatic test generation can be enabled via the `setup` command by responding with any variation of `Yes`, `yes`, `Y` or `y`, when asked...

`Would you like to auto-generate tests?`.

This closes Issue #5

### CODE REVIEW

File: bin/app.js
Comment (tdfranklin) [inline bin/app.js]: Rather than checking the entire conf file, I would just use ConfigStore's built in .get method.

https://www.npmjs.com/package/configstore

File: lib/create-test.js
Comment (tdfranklin) [inline lib/create-test.js]: There is a lot of copy/paste from the createComponent function... let's try to go to a higher level and pass some kind of logic to it to decide if a test or component should be passed... Alternatively, we could split out the actual creation of a file (all the fs parts) into it's own helper function then it could be shared between createComponent() and createTest() so we don't have so much repeated code.

File: lib/template-factory.js
Comment (tdfranklin) [inline lib/template-factory.js:21]: Love the way this looks!  Changing it to a switch statement will make it sooo much easier to expand on as we add further templates.  Great job!

File: lib/template-factory.js
Comment (tdfranklin) [inline lib/template-factory.js:16]: In other parts of the app, rather than throw error's, I've been preferring to just console.error a message and then exit gracefully.  I'm not sold on that convention, necessarily, but curious what your thoughts are on that?

**Review by tdfranklin — CHANGES_REQUESTED** [review]
Overall, looking very good and I think it's close.  One big issue, though, we may want to discuss.  Currently, we're assuming tests will be created in the same folder as the component.  In your experience, has that been a common convention for React projects?

Which I guess brings me to the next question, in order to keep Newton's Apple as unopinionated as possible, should we have some option where a user can store the path to their test folder for tests to be created?  Or should we just save the test w/ the component and rely on the user to move the file if they want it saved elsewhere?

File: bin/app.js
Comment (stevenpeniche) [reply bin/app.js]: Valid! I'll make that adjustment.

File: lib/create-test.js
Comment (stevenpeniche) [reply lib/create-test.js]: You're right! This is a great opportunity to refactor and make our functions more reusable.

File: lib/template-factory.js
Comment (stevenpeniche) [reply lib/template-factory.js:16]: I prefer throwing Errors for the simple fact that it facilitates deeper testing. As far as I'm aware there's no way to test if an Error is logged to the console other then by manually looking at the console. As long as the test assertions are working correctly and they pass, this Error should never be thrown in production so it's more of an Error to alert us that something went wrong. Let me know what you think?

**Comment by stevenpeniche** [comment]
As far as the bigger issue is concerned... I thought about going the route of allowing a user to specify a test folder path but we would run into problems with that. For instance, we wouldn't be able to properly `import` the corresponding component since we wouldn't know where it's located. We could ask the user for a components directory path but if the user decides to change up their project structure and split their components into corresponding **containers** and **components** directories, which is becoming increasingly common, we would need access to another path. 

Possible solution: One thing we can do is give the user an option to enable the creation of a test folder in the current directory and have all future test be added there if `-t` is passed or the auto test generation feature is on.

File: lib/template-factory.js
Comment (tdfranklin) [reply lib/template-factory.js:16]: I certainly understand and don't disagree with your reasoning.  Good food for thought, so I'll certainly keep that in mind!  Thank you.

**Comment by tdfranklin** [comment]
Yes, that's actually exactly what I've been trying to solve with the Redux stuff as well.  We're going to run into this exact same issue when implementing that, so I suppose let's take a minute and figure out how we want to resolve this.....

So, I see two potential options (and please let me know if you think of anything I forgot):

1. We make Newton's Apple a little opinionated and just create extra files in the same location as the component and rely on the user to reorganize as they see fit.

2. We find a way to keep Newton's Apple unopinionated by allowing the user to specify where files are created in a simple, unobtrusive way.

I think we can both agree that option 2 is preferable, if we can find a good way to implement it.  Here's the best idea I can come up with, so please let me know what you think or if you have a better idea...

What we re-engineer the `napp setup` command and maybe move the lifecycle method selection to another command (or perhaps gate it within the setup command) and then we require the user the run napp setup in the base project folder (so we capture the CWD and save it to the config as the base project directory), then we ask for the path (from the base project dir) where tests, reducers, dispatchers, store, etc is going to be saved and save that to a config, then we should be able to use that information to provide the proper import's on files to connect components to tests and everything.  What do you think?

**Comment by stevenpeniche** [comment]
I like that idea a lot! It'll add some required initial configuration but will make file generation trivial while maintaining the desired project structure. I'm guessing we would also ask for the path to the components directory in order to properly import from it?

**Comment by tdfranklin** [comment]
Yes!  However, we need to be careful to not have a components directory be mandatory.  I'd imagine some users may want to use NAPP for nothing more than simple component creation and not worry about the other generator stuff, so I want to maintain the current functionality as well where you can pass in the path if preferred. 

I guess maybe the best way to implement that is to have a componentPath in the config file and initially set it to null... then we can say if it's null, use the current logic on where to create and if there's a path there, use that instead.  Does that make sense?

**Comment by stevenpeniche** [comment]
That makes complete sense. Just one more question then... When no directory is passed we would be assuming the components are being created in the CWD that was captured on initial configuration?

**Comment by tdfranklin** [comment]
Yes, either in the CWD or the path from the CWD if a path is passed...

For instance, if componentPath in config is null:

```
$ napp new myComponent
```

This would create in currentDirectory/myComponent.js

```
$ napp new dirOne/dirTwo/myComponent
```

This would create in currentDirectory/dirOne/dirTwo/myComponent.js

However, let's say that in the config file you have projectPath and componentPath set... instead, the intended behavior would look like the following:

```
$ napp new myComponent
```

This would create in projectPath/componentPath/myComponent.js

```
$ napp new dirOne/dirTwo/myComponent
```

This would create in projectPath/componentPath/dirOne/dirTwo/myComponent.js

**Comment by stevenpeniche** [comment]
I like it! I think we should move forward with that implementation. Should I finish up the auto test generator feature before you begin work on re-engineering the initialization process or wait for you to implement that and rebase/rework my implementation based on the new process?

**Comment by tdfranklin** [comment]
Now, that said, there is already a "bug" with this idea that I haven't thought of a good way to solve yet...

Hopefully we can figure out a way to fix this before going live with the feature... the config file for Newton's Apple is global.  So if a user set's a path for files, then they go to work on another project and they want to use Newton's Apple for that project as well, they would have to change all the config paths or they would continue to point to the other project.  This would just make NAPP pain to use on multiple projects at the same time (which I'm hoping to avoid if possible).  

The only ideas I have thus far to solve this issue is to somehow have a separate config file for each project (but I haven't seen any options yet on how to do that with ConfigStore) or when creating the initial config, somehow nest an object w/ the project name first then all the individual settings inside of it... 

In any case, I'm still researching and trying to figure out how best to solve the issue, but wanted to bring it to your attention in case you have any ideas.

**Comment by tdfranklin** [comment]
> I like it! I think we should move forward with that implementation. Should I finish up the auto test generator feature before you begin work on re-engineering the initialization process or wait for you to implement that and rebase/rework my implementation based on the new process?

Good question... I guess for now, let's hold off until we figure out how to we want to approach the issue and set up the config.  We can just leave this PR open in the meantime and once we resolve that, finish up this work and merge it in.

**Comment by stevenpeniche** [comment]
>Now, that said, there is already a "bug" with this idea that I haven't thought of a good way to solve yet...

Other than creating an `napp.config` file in the project root directory and only falling back to the global configuration if one isn't found, the only other solutions I can think up are either nesting the configurations like you said or...

Upon executing the setup command, we could ask the user if they would like to set up/edit the global configuration file or a project specific configuration file. If it's global we would default to creating a `napp-config`. If it's project specific we would have the user pass in a name for the configuration file and store it with that instead of `napp-config`. We could run a check to see if it already exists before creating it and alert the user if one already exists with that name, then prompt them to enter another name. Following this process we could create a new command or a prompt within the setup command to switch between available configurations. We could also allow them to view/edit/delete the available configurations via that new command/prompt.

**Comment by tdfranklin** [comment]
Hrm, very interesting!  Yea, I "think" that should be possible with ConfigStore... I'll have to play around with it and see if I can get something like that working... Thanks for the great idea!

**Comment by tdfranklin** [comment]
Rather than comment in-line, I'll just post some thoughts here....

I don't think writeFile should be a part of the helper.js file, only because it has to do w/ file creation (which is "back-end" work) as opposed to "front-end" stuff.  So my thought would be move the writeFile() fuction to where the create-component file used to be and then you can just rename the file to either write-file.js or maybe we could name it helper.js if we think there's a possibility of the need for future "helper" functions needing to be created.

That said, I LOVE your create-file.js file.... I love the flow of having all the "create" functions there to determine if it's a component/test/etc file and I think it has great potential to continue to expand into the Redux phase with Reducers/Dispatchers/etc.

Do you agree with me or what are your thoughts?

**Comment by stevenpeniche** [comment]
I agree! I was looking at helper.js as more of an app wide helper file but I like the idea of creating one for the "backend" as well.

Glad you liked it 👍🏽

File: bin/parent/child/components/UglyButton.js
Comment (tdfranklin) [inline bin/parent/child/components/UglyButton.js]: What is this file for?  Looks like a test file got pushed by mistake?

File: bin/questions.js
Comment (tdfranklin) [inline bin/questions.js]: And this?

File: bin/questions.js
Comment (tdfranklin) [inline bin/questions.js]: I would move this further down the path.... I think the first question we want to ask is if the cwd is the correct directory and then set up a project name.... probably move this close to the  bottom, below the component dir question?

File: lib/create-file.js
Comment (tdfranklin) [inline lib/create-file.js]: I think a lot of this is my own fault, but we may want to try and find some better names for variables here... 

What do you think about going forward, rather than writing comments to try and explain what something is or what it's doing, let's try to create really robust names that are self explanatory as to what that variable is referencing?  I'm just trying to look at it from the perspective of someone who didn't know the code at all.... so maybe we can try to keep that in mind going forward to be a little more verbose (and again, that's mostly my fault... I'm really bad about choosing very generic variable names, lol).

File: lib/create-file.js
Comment (tdfranklin) [inline lib/create-file.js]: So before we merge this in, trying to think towards the future... we're going to be looking at probably at least 1 or 2 more "file types"... do you think the best approach is to have a function for each type (eg; createComponent, createTest, createReducer, createStore, etc) or should we try to merge it into bigger function to to keep it a little DRYer?

On one hand, I think having a separate function for each makes the code a bit more readable and adhere's more to separation of concern's, but there's also a lot of repetitive code, so it makes it less DRY... What are your thoughts?  Maybe we should find some middle ground and have almost another factory-type setup to determine the template and filename?  I'm really not sure, lol.

File: lib/helpers.js
Comment (tdfranklin) [inline lib/helpers.js]: Wow, quite the bit of code here!  Honestly impressed!  Just a thought, though, but couldn't you just use path.relative to make your life easier?  Just something like path.relative(componentDir, testDir)?

**Review by tdfranklin — COMMENTED** [review]
Looking great so far!  Thank you again for all your hard work!

File: bin/parent/child/components/UglyButton.js
Comment (stevenpeniche) [reply bin/parent/child/components/UglyButton.js]: Sneaky test files always find a way to slip through... I'll remove that.

File: bin/questions.js
Comment (stevenpeniche) [reply bin/questions.js]: What's wrong here?

File: bin/questions.js
Comment (stevenpeniche) [reply bin/questions.js]: It flows better, you're right.

File: lib/create-file.js
Comment (stevenpeniche) [reply lib/create-file.js]: lol, I'll edit these variable names and keep that in mind moving forward.

File: lib/create-file.js
Comment (stevenpeniche) [reply lib/create-file.js]: I think the best approach is to have a function for each file type for the exact reason you mentioned. It allows for better readability and maintainability. Another positive is we're able to tweak each function in isolation instead of affecting all the creation of all file types at once.

What I can do is extrapolate the formatting of the `filePath` into its own function and place it in our helpers object.

File: lib/helpers.js
Comment (stevenpeniche) [reply lib/helpers.js]: Didn't even know `path.relative()` existed 😅. I'll refactor it to that.

File: bin/parent/child/components/UglyButton.js
Comment (tdfranklin) [reply bin/parent/child/components/UglyButton.js]: Lol, no worries!

File: bin/questions.js
Comment (tdfranklin) [reply bin/questions.js]: Ignore me here... can't find any issues now, so guess I was just tired last night, lol.  Sorry about that!

File: lib/create-file.js
Comment (tdfranklin) [reply lib/create-file.js]: Yea, I'm gonna try to get better at that as well... I'm hoping at some point that more people will join in to contribute, so I need to do a better job of keeping that mind as I code.  Bad habits are hard to break, though, lol!

File: lib/create-file.js
Comment (tdfranklin) [reply lib/create-file.js]: Yea, after sleeping on it, I think I agree with you.  We may repeat a bit of code, but I think it does keep it much more readable and easier to work on.  Thanks for keeping me honest!

File: lib/helpers.js
Comment (tdfranklin) [reply lib/helpers.js]: Lol, no worries!  I don't think it's something that you would commonly use.  Just hate you spent all that time pretty much building it yourself, but I'm sure some of that code will come in handy 🥇

File: lib/create-file.js
Comment (stevenpeniche) [reply lib/create-file.js]: Funny thing... After refactoring the formatting of the `filePath` I saw an opportunity to convert `createFIle` into the factory-type function you mentioned. This will definitely make things cleaner while still allowing us to add `fileType` specific behavior in the future! Check out the new commit and let me know what you think.

**Review by tdfranklin — APPROVED** [review]
Yes, yes, and more yes!!  Excellent work!  I think that is a great way to do it and leaves plenty of room for expansion!

File: bin/app.js
Comment (tdfranklin) [inline bin/app.js]: Thinking out loud here, so bare with me....
I'm debating weather or not it's a good idea to have this extra logic in the action itself for test creation... I know it's a minor thing, but in my head it seems more like a "back end" type job.... I'm actually thinking maybe we should remove the `changeAllSettings()` calls as well... what do you think?

File: bin/app.js
Comment (tdfranklin) [inline bin/app.js]: Not sure if this was auto formatted, but a little odd to move this to one line when the other's have each method call on it's on line... either way not a big deal, just thought I'd mention it.

File: bin/configstore.js
Comment (tdfranklin) [inline bin/configstore.js:74]: I may be mistaken, but I'm pretty sure `ifPathExists()` is just a helper function for `setupProject()` right?  If so, I don't think there's any reason to export it.  I may just be missing it being used somewhere else, though?

File: lib/helpers.js
Comment (tdfranklin) [inline lib/helpers.js]: Again, nothing to block on here, but just curious why export as helpers instead of the format we have in the other files?  They both 100% amount to the same thing, but I think we should decide on a convention and keep it consistent across the app.  I'm perfectly fine with either way you want to handle it, though.

**Review by tdfranklin — COMMENTED** [review]
Tests look great!  Quite impressive!  Reading through I had a few thoughts for you to review and see what you think.  Just let me know or feel free to make any changes you think would be good and then merge when you feel ready.  I'm really excited!!

File: bin/app.js
Comment (stevenpeniche) [reply bin/app.js]: I like the idea of extracting the calls to` changeAllSettings()` as well as the `autoGenerateTests` check and placing it in **process-config.js**. In that case, all that will remain is the single call to `processConfig()`.

File: bin/app.js
Comment (stevenpeniche) [reply bin/app.js]: This was definitely an auto format thing, I'll edit it.

File: bin/configstore.js
Comment (stevenpeniche) [reply bin/configstore.js:74]: You're completely correct. The only reason I'm exporting it is for testing purposes.

File: lib/helpers.js
Comment (stevenpeniche) [reply lib/helpers.js]: I agree with keeping consistent with the established convention. I'll switch this over.

**Pull request merged**
